src/Conversion/ONNXToKrnl/Tensor/Gather.cpp

```diff
@@ -0,0 +1,117 @@
+//===----------------Gather.cpp - Lowering Gather Op----------------------=== //
+//
+// Copyright 2020 The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file lowers the ONNX Gather Operator to Krnl dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "src/Conversion/ONNXToKrnl/ONNXToKrnlCommon.hpp"
+
+using namespace mlir;
+
+struct ONNXGatherOpLowering : public ConversionPattern {
+  ONNXGatherOpLowering(MLIRContext *ctx)
+      : ConversionPattern(mlir::ONNXGatherOp::getOperationName(), 1, ctx) {}
+
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    ONNXGatherOpAdaptor operandAdaptor(operands);
+    ONNXGatherOp gatherOp = llvm::dyn_cast<ONNXGatherOp>(op);
+    auto loc = op->getLoc();
+    // get input operands, shapes, and rank
+    Value data = operandAdaptor.data();
+    auto dataShape = data.getType().cast<MemRefType>().getShape();
+    int64_t dataRank = dataShape.size();
+    Value indices = operandAdaptor.indices();
+    auto indicesShape = indices.getType().cast<MemRefType>().getShape();
+    int64_t indicesRank = indicesShape.size();
+    int64_t axisIndex = gatherOp.axis().getSExtValue();
+    // get output info
+    auto outputMemRefType = convertToMemRefType(*op->result_type_begin());
```

Comment:
Question not specific to this PR: Is there a pass to translate all tensors, including inputs and outputs,  to memref?

Reply:
May be there should be two ONNX dialects, ONNX on tensors and ONNX on buffers but unfortunately it's not there.

Reply:
My understanding is that ONNX is for tensors, and krnl is for memrefs, and that the translations occurs during the lowering. If there is a need for ONNX-level transformation that can only done with memrefs, then we should think about this.

Reply:
I mean, if we have ONNX on memrefs then we can directly move from ONNX on memrefs to affine or linalg etc. We can even use stuff from tensorflow's lmhlo dialect. If there is a better way do suggest.